src/h3/_cy/cells.pyx

```diff
@@ -0,0 +1,281 @@
+
+
+cpdef H3int parent(H3int h, res=None) except 1:
+    check_cell(h)
+
+    if res is None:
+        res = resolution(h) - 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    check_res(res)
+
+    return h3lib.h3ToParent(h, res)
+
+
+cpdef int distance(H3int h1, H3int h2) except -1:
+    """ compute the hex-distance between two hexagons
+    """
+    check_cell(h1)
+    check_cell(h2)
+
+    d = h3lib.h3Distance(h1,h2)
+
+    return d
+
+cpdef H3int[:] disk(H3int h, int k):
+    """ Return cells at grid distance `<= k` from `h`.
+    """
+    check_cell(h)
+    if k < 0:
+        raise H3ValueError('Invalid ring size: {}'.format(k))
+
+    n = h3lib.maxKringSize(k)
+
+    ptr = create_ptr(n) # todo: return a "smart" pointer that knows its length?
+    h3lib.kRing(h, k, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+
+cpdef H3int[:] _ring_fallback(H3int h, int k):
+    """
+    `ring` tries to call `h3lib.hexRing` first; if that fails, we call
+    this function, which relies on `h3lib.kRingDistances`.
+
+    Failures for `h3lib.hexRing` happen when that alg runs into a pentagon.
+    """
+    check_cell(h)
+
+    n = h3lib.maxKringSize(k)
+    # array of h3 cells
+    ptr = create_ptr(n)
+
+    # array of cell distances from `h`
+    dist_ptr = <int*> stdlib.calloc(n, sizeof(int))
+    if dist_ptr is NULL:
+        raise MemoryError()
+
+    h3lib.kRingDistances(h, k, ptr, dist_ptr)
+
+    distances = <int[:n]> dist_ptr
+    distances.callback_free_data = stdlib.free
+
+    for i,v in enumerate(distances):
+        if v != k:
+            ptr[i] = 0
+
+    mv = create_mv(ptr, n)
+
+    return mv
+
+cpdef H3int[:] ring(H3int h, int k):
+    """ Return cells at grid distance `== k` from `h`.
+    Collection is "hollow" for k >= 1.
+    """
+    check_cell(h)
+
+    n = 6*k if k > 0 else 1
+    ptr = create_ptr(n)
+
+    flag = h3lib.hexRing(h, k, ptr)
+
+    # if we drop into the failure state, we might be tempted to not create
+    # this mv, but creating the mv is exactly what guarantees that we'll free
+    # the memory. context manager would be better here, if we can figure out
+    # how to do that
+    mv = create_mv(ptr, n)
+
+    if flag != 0:
+        mv = _ring_fallback(h, k)
+
+    return mv
+
+
+
+cpdef H3int[:] children(H3int h, res=None):
+    check_cell(h)
+
+    if res is None:
+        res = resolution(h) + 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    check_res(res)
+
+    n = h3lib.maxH3ToChildrenSize(h, res)
+
+    ptr = create_ptr(n)
+    h3lib.h3ToChildren(h, res, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+
+
+cpdef H3int[:] compact(const H3int[:] hu):
+    # todo: the Clib can handle 0-len arrays because it **avoids**
+    # dereferencing the pointer, but Cython's syntax of
+    # `&hu[0]` **requires** a dereference. For Cython, checking for array
```

Comment:
I'm still surprised this requires a dereference, in C it would be written the same but I believe not actually cause a dereference.

Reply:
I agree this feels awkward and unnecessary, but I couldn't find a better way around it earlier. Let's kick this to an issue and try to figure it out later?

Reply:
issue: https://github.com/uber/h3-py/issues/105